File: dotnet/src/Microsoft.Agents.Validation/Validators/MessageValidators/AgentMessageValidator.cs
using Microsoft.Agents.Abstractions.Models;
using System.Text.RegularExpressions;

namespace Microsoft.Agents.Validation.Validators.MessageValidators;

/// <summary>
/// Validator for AgentMessage.
/// </summary>
public partial class AgentMessageValidator : MessageValidatorBase<AgentMessage>
{
    [GeneratedRegex(@"^run_[a-zA-Z0-9_-]+$")]
    private static partial Regex CompletionIdRegex();

    protected override List<ValidationError> ValidateSpecificFields(AgentMessage message, ValidationContext? context)
    {
        var errors = new List<ValidationError>();

        // MSG-011: AgentMessage must have at least one content item
        if (message.Contents == null || message.Contents.Count == 0)
        {
            errors.Add(CreateError(
                ValidationErrorCode.MSG_011,
                "AgentMessage must have at least one content item",
                "Contents"));
        }
        else
        {
            // ROLE-003: AgentMessage can contain FunctionCallContent, TextContent, TextReasoningContent
            // ROLE-007: FunctionCallContent can only appear in AgentMessage
            // These are validated at content level

            // Validate content items
            errors.AddRange(ValidateContentItems(message.Contents, context));
        }

        // REL-009: AgentMessage completion-id must be valid format (run_*)
        if (!string.IsNullOrWhiteSpace(message.CompletionId) &&
            !CompletionIdRegex().IsMatch(message.CompletionId))
        {
            errors.Add(CreateError(
                ValidationErrorCode.REL_009,
                $"AgentMessage completion-id '{message.CompletionId}' must match format 'run_*'",
                "CompletionId"));
        }

        // REL-010: AgentMessage agent-id must be non-empty when present
        if (message.AgentId != null && string.IsNullOrWhiteSpace(message.AgentId))
        {
            errors.Add(CreateError(
                ValidationErrorCode.REL_010,
                "AgentMessage agent-id must be non-empty when present",
                "AgentId"));
        }

        return errors;
    }
}

File: typescript/packages/agents-protocol-validation/src/validators/message/AgentMessageValidator.ts
import type { AgentMessage } from '@microsoft/agents-protocol-abstractions';
import type { ValidationContext } from '../../core/ValidationContext';
import { MessageValidatorBase } from '../base/MessageValidatorBase';
import { ValidationError } from '../../core/ValidationError';
import { ValidationErrorCode } from '../../core/ValidationErrorCode';

/**
 * Validator for AgentMessage.
 */
export class AgentMessageValidator extends MessageValidatorBase<AgentMessage> {
  protected validateSpecificFields(message: AgentMessage, context?: ValidationContext): ValidationError[] {
    const errors: ValidationError[] = [];

    // MSG-011: Agent messages must have contents (checked in base class)

    // ROLE-007: At most one text-reasoning content
    if (message.contents) {
      const reasoningCount = message.contents.filter((c) => c.kind === 'textReasoning').length;
      if (reasoningCount > 1) {
        errors.push(
          new ValidationError(
            ValidationErrorCode.ROLE_007,
            'AgentMessage can have at most one text-reasoning content',
            'contents'
          )
        );
      }

      // BIZ-001: If text-reasoning exists, it must come before other content
      const reasoningIndex = message.contents.findIndex((c) => c.kind === 'textReasoning');
      if (reasoningIndex > 0) {
        errors.push(
          new ValidationError(
            ValidationErrorCode.BIZ_001,
            'text-reasoning content must come before other content types',
            'contents'
          )
        );
      }
    }

    // REL-009, REL-010: Validate function calls (if context available)
    // These are validated at content level and by relationship validator

    return errors;
  }

  protected getAllowedContentTypes(): string[] {
    // ROLE-003: Agent messages can contain text, functionCall, textReasoning, refusal, error
    return ['text', 'functionCall', 'textReasoning', 'refusal', 'error'];
  }

  protected getRoleSpecificErrorCode(): string {
    return ValidationErrorCode.ROLE_003;
  }

  protected getMessageTypeName(): string {
    return 'AgentMessage';
  }

  protected requiresContent(): boolean {
    return true;
  }
}
